### PR TITLE
Subworkflow to transfer BIDS `.json` to `nextflow channels`

### PR DESCRIPTION
**Pull request by gagnonanthony**
New subworkflow to cover the gap between the `json` file outputted by `IO_READBIDS` and usable `nextflow channels`. It uses all currently collected files from `scil_bids_validate.py`, but should be adapted in future pipelines to meet the specific file requirements.

**Tests are not super informative since there aren't actual files in the test data.**

### CODE REVIEW

File: subworkflows/nf-neuro/io_bids/main.nf [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+                                "structure using the bids-validator tool:" +
+                                "https://bids-standard.github.io/bids-validator/" +
+                                "or the docker version: " +
+                                "https://hub.docker.com/r/bids/validator"
+                    }
+                }
+
+                // ** Collecting the files ** //
+                return [
+                    [id: sid, session: ses, run: run, tr: tr, phase: phase, revphase: revphase],
```
Comment (AlexVCaron) [inline subworkflows/nf-neuro/io_bids/main.nf:55]: I'd like it, but I think having parameters like TR that vary between modalities will break channel jointing.

File: subworkflows/nf-neuro/io_bids/main.nf [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        fs_folder,
+        bidsignore
+    )
+    ch_versions = ch_versions.mix(IO_READBIDS.out.versions)
+
+    // ** Converting the json file into channels. ** //
+    ch_files = IO_READBIDS.out.bidsstructure
+        .flatMap{ layout ->
+            def json = new groovy.json.JsonSlurper().parseText(layout.getText())
+            json.collect { item ->
```
Comment (AlexVCaron) [inline subworkflows/nf-neuro/io_bids/main.nf:26]: Just to clarify this statement, add a `return` before. I know collect does it automatically, but I think it could mislead the reader.

File: subworkflows/nf-neuro/io_bids/main.nf (hunk 1 above)
Comment (AlexVCaron) [reply subworkflows/nf-neuro/io_bids/main.nf:55]: I would be fine with adding an extra map to the outputs of IO_BIDS. There isn't a way to process it for us, but for now we can garbage it and start updating the modules that could use it.

**Review by AlexVCaron — CHANGES_REQUESTED** [review]
Very nice ! Some fine tuning to do and considerations. It will be super useful in its final state !

**Review by AlexVCaron — CHANGES_REQUESTED** [review]
After thoughts, I'm going to wait for @arnaudbore comments on this. I think I'd like a `multi-bids` test before further reviewing. The way it's implemented, we can pass more than one **bids database** and they'll get flatten out into a merged unified product. This way, we can merge something like a **DWI bids** with an **anat bids** or anything else (thinking about `fs`, current use-case that would need to be implemented in nf-tractoflow).

It's awesome, the start of what I was envisioning for this subworkflow, but edge cases are not covered with the current test's state, and I realise there is **a lot of edges**.  Lets sync so I can upload data that works for this.

However, to control those edges, I think we should add another layer. To control those edges better, with the bonus of making the subworkflow even more useful. I'd add a `database name` as a meta tag (to all 3 inputs), on which we can start `collecting` associated `bids datasets`. Then we `merge` them, and afterwards `flatten` them to obtain the `channel inputs`.

**Comment by gagnonanthony** [comment]
I understand the point here, but from my understanding, only a single folder input is supported here? (`IO_READBIDS` supports only a single folder and returns a single JSON, which will be parsed by the workflow afterward). 

Also, `freesurfer` folder input is already supported (at least for the `wmparc` and `aseg-aparc` files), I agree it is not a full support, but the minimum needed for `tractoflow` is here.

I do agree that we currently do not test anything at all, so we should definitely get on that. However, I'm not sure we should support the `multi-bids` option. I guess if both bids folders have the same acquisition parameters, it would work. However, from my own experience, it is rarely the case, and I need to tweak some parameters when executing a pipeline to fit the input data. In the case where we have, let's say, three datasets with parameters that need to be tuned for each, it wouldn't work in our current implementation of our pipelines.

Those are just my thoughts @AlexVCaron. Happy to discuss it further! 😄

File: subworkflows/nf-neuro/io_bids/main.nf [hunk 3]
```diff
@@ -5,12 +5,20 @@ workflow IO_BIDS {
...
 
     ch_versions = Channel.empty()
 
+    // ** Sanity check to ensure channels are single-item ** //
+    BIDS_FOLDER_SANITY_CHECK = bids_folder.collect().map { folders ->
+        if (folders.size() > 1) {
+            error "ERROR: You must supply only a single BIDS folder."
+        }
+        return folders[0]
+    }
```
[the remark this answers is not in the document]
Comment (gagnonanthony) [reply subworkflows/nf-neuro/io_bids/main.nf]: You are right! I removed it

**Comment by AlexVCaron** [comment]
LGTM whenever you guys are happy with it ! @arnaudbore

**Review by AlexVCaron — APPROVED** [review]

**Review by arnaudbore — APPROVED** [review]
GTG
